pkg-manager/resolve-dependencies/src/index.ts

```diff
@@ -319,20 +321,87 @@ function verifyPatches (
+    size: number
+    type: 'File'
+    path: string
+    contentType: string
+    hex: string
+    isBinary: boolean
+    linesCount: number
+  }
+
+  interface RegistryFileList {
+    files: {
+      [path: string]: RegistryFileFields
+    }
+  }
+
+  let registryUrls = ['registry.npmjs.org/', 'registry.npmjs.com/']
+  if (Array.isArray(useExperimentalNpmjsFilesIndex)) {
+    registryUrls = useExperimentalNpmjsFilesIndex
+  }
+  // check if any of the registryUrls are contained in the pkgUrl
+  const registryUrl = registryUrls.find((url) => pkgUrl.startsWith(url))
+  if (!registryUrl) {
+    globalWarn(`${pkgName}@${pkgVer} (${pkgUrl}) is not in NPMJS Files Index URL Allow List: Fallback to requiresBuild=true`)
+    return true
+  }
+
+  const url = `https://npmjs.com/package/${pkgName}/v/${pkgVer}/index`
+  try {
+    const fetchResult = await fetch(url)
+    if (!fetchResult.ok) {
+      globalWarn(`Unable to  fetch npmjs file list for ${pkgName}@${pkgVer}: Fallback to requiresBuild=true`)
+      return true
+    }
+
+    const fetchJSON = await fetchResult.json() as RegistryFileList
+    if (fetchJSON.files == null) {
+      globalWarn(`Unable to parse npmjs file list for ${pkgName}@${pkgVer}: Fallback to requiresBuild=true`)
+      return true
+    }
+    const regFS = fetchJSON.files
+    if (regFS['/binding.gyp'] || regFS['/hooks'] || regFS['/hooks/']) {
+      return true
+    }
+    const pkgJsonHex = regFS['/package.json'].hex
+  const pkgJson = await (await fetch(`https://npmjs.com/package/${pkgName}/file/${pkgJsonHex}`)).json() as any // eslint-disable-line
+    if (pkgJson?.scripts != null && (
+      Boolean(pkgJson.scripts.preinstall) ||
+      Boolean(pkgJson.scripts.install) ||
+      Boolean(pkgJson.scripts.postinstall)
+    )) {
+      return true
+    }
+    return false
+  } catch (e) {
+    if (e instanceof Error) {
+      throw new PnpmError(`Failed to fetch ${pkgName}@${pkgVer}`, e.message)
+    }
+  }
+  return false
+}
+
 async function finishLockfileUpdates (
   dependenciesGraph: DependenciesGraph,
   pendingRequiresBuilds: string[],
-  newLockfile: Lockfile
+  newLockfile: Lockfile,
+  useExperimentalNpmjsFilesIndex?: boolean | string[]
 ) {
   return Promise.all(pendingRequiresBuilds.map(async (depPath) => {
     const depNode = dependenciesGraph[depPath]
     if (!depNode) return
     try {
       let requiresBuild!: boolean
       if (depNode.optional) {
-        // We assume that all optional dependencies have to be built.
-        // Optional dependencies are not always downloaded, so there is no way to know whether they need to be built or not.
-        requiresBuild = true
+        if (useExperimentalNpmjsFilesIndex) {
+        // Attempt to use NPM's file list to determine if the package requiresBuild
+          requiresBuild = await fetchBuildFromRegistryFS(depNode.name, depNode.version, depNode.id, useExperimentalNpmjsFilesIndex)
+        } else {
+          requiresBuild = true
+        }
```

Comment:
we're going to do this for non-optional packages when lockfile-only installation is happening and the setting is turned on. Optional deps should be always true for now.

Reply:
This change should be reverted. Below, in the non-optional else statement should experimental npmjs files index be used.